src/asio_service.cxx

```diff
@@ -775,29 +775,35 @@ class asio_rpc_listener
 
 public:
     virtual void stop() override {
+        auto_lock(listener_lock_);
         stopped_ = true;
         acceptor_.close();
     }
 
     virtual void listen(ptr<msg_handler>& handler) override {
+        std::lock_guard<std::mutex> guard(listener_lock_);
         handler_ = handler;
         stopped_ = false;
-        start();
+        start(guard);
     }
 
     virtual void shutdown() override {
-        auto_lock(session_lock_);
-        for (auto& entry: active_sessions_) {
-            ptr<rpc_session> s = entry;
-            s->stop();
-            s.reset();
+        {
+            auto_lock(session_lock_);
+            for (auto& entry: active_sessions_) {
+                ptr<rpc_session> s = entry;
+                s->stop();
+                s.reset();
+            }
+            active_sessions_.clear();
         }
-        active_sessions_.clear();
+
+        auto_lock(listener_lock_);
         handler_.reset();
     }
 
 private:
-    void start() {
+    void start(std::lock_guard<std::mutex> & listener_lock_) {
```

Comment:
Looks like this parameter (`std::lock_guard<std::mutex> & listener_lock_`) is not used, hence can be removed?

Reply:
this is just a convention to make sure we lock before calling start

Reply:
Ok, then can you please rename it without `_` suffix (`listener_lock`)? as it is a local parameter.

Reply:
ah yes, can do that